[Cheong Yee Ming] iP

CheongYeeMing opened the pull request:
# Duke

> "The way to get started is to quit talking and begin doing." -Walt Disney ([source](https://blog.hubspot.com/sales/famous-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is

1. download it from [here](https://github.com/CheongYeeMing/ip/releases).
2. double-click it.
3. add your tasks.
4.  let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

CheongYeeMing committed:
Level-1

CheongYeeMing committed:
Level-2

CheongYeeMing committed:
Level-3

CheongYeeMing committed:
Level-4

CheongYeeMing committed:
A-TextUiTesting

CheongYeeMing committed:
Level-5

CheongYeeMing committed:
Level-6

CheongYeeMing committed:
Level-7

CheongYeeMing committed:
Level-8

CheongYeeMing committed:
Level-8

CheongYeeMing committed:
A-MoreOOP

CheongYeeMing committed:
A-Packages

CheongYeeMing committed:
A-JUnit

CheongYeeMing committed:
A-JavaDoc

CheongYeeMing committed:
branch-A-CodingStandard

CheongYeeMing committed:
Level-9

pualixue commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /**
+     *
```
Is the method summary missing?

pualixue commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,98 @@
+     * Saves tasks in TaskList to local directory.
+     * Folder and file will be created if not already present.
+     *
+     * @param taskList
```
Is the description of `taskList` missing?

pualixue commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     * @param task Task to be added into task list.
+     * @param taskLeft Total number of tasks in the task list.
+     */
+    public void taskAddedMessage(Task task, int taskLeft) {
```
Perhaps this method name can be changed to a verb (i.e. `printTaskAddedMessage`)?
Similarly for some of the methods below.

pualixue commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Constructor for an AddCommand.
+     *
+     * @param taskList Handles all operations regarding tasks.
```
I like how the descriptions have the same indentation which makes it easy to read. 👍

pualixue reviewed (commented):
LGTM. Just some nits to fix.

whoisjustinngo commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,38 @@
+     *
+     * @throws DukeException when an error occurs.
+     */
+    public abstract void runCommand() throws DukeException;
```
Maybe you can consider renaming the method `execute` or simply `run` to clearly distinguish _the method to run the command_ from _the type of command_ `(i.e. AddCommand, DeleteCommand etc.)`

whoisjustinngo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+     * @throws DukeException If user input is invalid.
+     */
+    public Command parseUserInput(String userInput) throws DukeException {
+        String[] splitUserInput = userInput.split(" ");
```
Consider using a noun like `userInputTokens` instead of `splitUserInput`

whoisjustinngo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+                if (byIndex <= 0) {
+                    throw new NoDateTimeException(ui);
+                }
+                String deadlineDetails = commandDetails.substring(0, byIndex);
+                if (deadlineDetails.isBlank()) {
+                    throw new NoTaskDescriptionException(ui);
+                }
+                try {
+                    LocalDateTime by = LocalDateTime.parse(commandDetails.substring(deadlineDetails.indexOf("by") + 3),
+                            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+                    return new Deadline(deadlineDetails, false, by);
+                } catch (DateTimeParseException e) {
+                    throw new InvalidDateTimeException(ui);
+                }
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new NoTaskDescriptionException(ui);
+            }
+
+
+        case EVENT:
+            try {
+                int atIndex = commandDetails.indexOf("/at") - 1 ;
+                if (atIndex <= 0) {
+                    throw new NoDateTimeException(ui);
+                }
+                String eventDetails = commandDetails.substring(0, atIndex);
+                if (eventDetails.isBlank()) {
+                    throw new NoTaskDescriptionException(ui);
+                }
+                try {
+                    LocalDateTime at = LocalDateTime.parse(commandDetails.substring(eventDetails.indexOf("at") + 3),
+                            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+                    return new Event(eventDetails, false, at);
+                } catch (DateTimeParseException e) {
+                    throw new InvalidDateTimeException(ui);
+                }
+            } catch (StringIndexOutOfBoundsException e) {
+                throw new NoTaskDescriptionException(ui);
+            }
+
```
The code in the `Deadline` and `Event` cases are very similar hence I think you could combine both into one case; alternatively, assuming all `taskType` inputs are valid, consider the `ToDo` case first and let the `Deadline` and `Event` be the default. To account for the 'by' and 'at' issue, maybe you can define an auxiliary delimiter string, if event then the string is 'at', else 'by'. This might help significantly reduce code duplication 👍🏼

whoisjustinngo commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> list;
```
Perhaps `tasks` would be more descriptive than `list`

whoisjustinngo commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        switch (commandType) {
+        case "list":
+            return new ListCommand(taskList, storage, ui);
+
+        case "bye":
+            return new ExitCommand(taskList, storage, ui);
+
+        case "deadline":
+            return new AddCommand(taskList, storage, ui, parseTaskInput(TaskType.DEADLINE, userInput));
+
+        case "event":
+            return new AddCommand(taskList, storage, ui, parseTaskInput(TaskType.EVENT, userInput));
+
+        case "todo":
+            return new AddCommand(taskList, storage, ui, parseTaskInput(TaskType.TODO, userInput));
+
+        case "done":
+            return new DoneCommand(taskList, storage, ui, Integer.parseInt(splitUserInput[1]));
+
+        case "delete":
+            return new DeleteCommand(taskList, storage, ui, Integer.parseInt(splitUserInput[1]));
+
+        case "find":
+            return new FindCommand(taskList, storage, ui, userInput.substring(userInput.indexOf(" ") + 1));
+
+        default:
+            throw new NoSuchCommandException(ui);
+        }
```
I like how clean and neat this is 👍🏼

whoisjustinngo reviewed (commented):
LGTM, just some minor cosmetic issues to consider 🥇

chengseong commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -5,14 +5,33 @@
 import duke.ui.Ui;
 
+/**
+ * Command to add a task to the task list.
```
I think it would be good to document down what is contained in each Command

chengseong commented on src/main/java/duke/command/Command.java:
```diff
@@ -5,16 +5,34 @@
     protected final Ui ui;
 
+    /**
+     * Constructor for a DukeCommand.
```
You mention a `DukeCommand` in the JavaDocs, but the constructor creates a `Command`, so it might be good to streamline this and standardize to 1 name for the `Command` class

chengseong commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -3,24 +3,58 @@
+ */
 public class Deadline extends Task {
 
     protected final LocalDateTime by;
```
A name like `by` is not really descriptive enough. Maybe consider something like `deadline`.

chengseong commented on src/main/java/duke/task/Event.java:
```diff
@@ -3,24 +3,58 @@
+ */
 public class Event extends Task {
 
     protected final LocalDateTime at;
```
Same as `by`,  `at` is not very descriptive

chengseong reviewed (commented):
LGTM! Just a few nits with the JavaDocs but generally looks great!

CheongYeeMing committed:
Merge remote-tracking branch 'origin/add-gradle-support'

CheongYeeMing committed:
A-Gradle

CheongYeeMing committed:
A-CheckStyle

CheongYeeMing committed:
Level-10

CheongYeeMing committed:
A-Assertions

No assertions used in code.

Assertions help detect programmer mistakes.

Adding assert statement to the following classes:
* `AddCommand`
* `DeleteCommand`
* `DoneCommand`
* `Parser`
* `Storage`
* `Deadline`
* `Task`
* `TaskList`

Assert statements are added in cases where `null` is passed into the respective `Task` constructors, detects empty tasks.

CheongYeeMing committed:
A-CodeQuality

Nested class.

Outer class now had 2 responsibilities. Breaches the Single Responsibility Principle.

Extracting nested class `ExitProgram` within `Main` class.

Avoid nesting of classes, adhere to Single Responsibility Principle.

CheongYeeMing committed:
A-Streams

For loops used for iteration.

Too many lines of code, does not look good.

Replace for loops with streams.

Benefits of streams:
* Write functions at more abstract level
* Make code more expressive
* Less verbose code
* More Intuitive code
* Less error prone code

CheongYeeMing committed:
Merge pull request #1 from CheongYeeMing/branch-A-Assertions

A-Assertions

CheongYeeMing committed:
Merge pull request #2 from CheongYeeMing/branch-A-CodeQuality

A-CodeQuality

CheongYeeMing committed:
Merge pull request #3 from CheongYeeMing/branch-A-Streams

A-Streams

CheongYeeMing committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/CheongYeeMing/ip

CheongYeeMing committed:
Merge branch 'branch-A-Streams' of https://github.com/CheongYeeMing/ip

CheongYeeMing committed:
Merge branch 'master' of https://github.com/CheongYeeMing/ip

CheongYeeMing committed:
BCD-Extension

CheongYeeMing committed:
A-BetterGui

CheongYeeMing committed:
A-UserGuide

CheongYeeMing committed:
A-Release

CheongYeeMing committed:
Merge branch 'master' of https://github.com/CheongYeeMing/ip

CheongYeeMing committed:
A-UserGuide
